typescript/src/game_loop/game-rxjs.ts

```diff
@@ -1,115 +1,158 @@
+  constructor(
+    public readonly message: Message,
+    public readonly from: BaseSocket
+  ) {}
 }
 
-function empty<T>(): Observable<T> {
-    return new Observable((observer: Subscriber<T>) => {
-        observer.complete();
+export function runRxJSLoop(
+  subscriber: Subscriber<GameResults>,
+  s1: RxSocket,
+  s2: RxSocket
+): void {
+  const stats = new GameStat();
+  const queue: MessageEnvelope[] = [];
+
+  const world = new GameWorld(s1, s2);
+  subscriber.add(() => world.stop());
+
+  function close(other: RxSocket) {
+    if (world.done) {
+      return;
+    }
+    other.push(errorGameOver("The other player disconnected"));
+    subscriber.next(new GameResultsImpl(stats, s1, s2, true));
+    subscriber.complete();
+  }
+
+  subscriber.add(
+    s1.events.subscribe({
+      next: (message) => queue.push(new MessageEnvelope(message, s1)),
+      complete: () => close(s2),
+    })
+  );
+
+  subscriber.add(
+    s2.events.subscribe({
+      next: (message) => queue.push(new MessageEnvelope(message, s2)),
+      complete: () => close(s1),
     })
+  );
+
+  subscriber.add(
+    getRxJSGameLoop(getTickRate()).subscribe((delta) => {
+      stats.addDelta(delta);
+
+      // 1. process messages
+      const messageEnvolopes = Array.from(queue);
+      queue.length = 0;
+      for (let i = 0; i < messageEnvolopes.length; i++) {
+        const m = messageEnvolopes[i];
+        world.processMessage(m.from, m.message);
+      }
+
+      // 2. update all positions
+      world.update(delta);
+
+      // 3. process collisions
+      world.collisions();
+
+      if (world.done) {
+        const gameResult = new GameResultsImpl(
+          stats,
+          world.getWinner(),
+          world.getLoser(),
+          false
+        );
+        subscriber.next(gameResult);
+        subscriber.complete();
+      }
+    })
+  );
 }
 
 export default function gameCreator(server: Server) {
-    server.on().pipe(
-        mergeMap(sockets => setupWithRxJS(sockets)),
-        tap(([s1, s2]) => {
-            s1.push(createMessage(MessageType.Play));
-            s2.push(createMessage(MessageType.Play));
-        }),
-        mergeMap(([s1, s2]) => {
-            GameStat.activeGames++;
-            return runRxJSLoop([s1, s2]).pipe(
-                tap({
-                    next: ([stats, winner, loser]) => {
-                        winner.push(createWinnerMessage(stats), () => winner.close());
-                        loser.push(createLoserMessage(), () => loser.close());
-                        GameStat.activeGames--;
-                    },
-                    error: (_) => {
-                        GameStat.activeGames--;
-                    }
-                }),
-                onErrorResumeNext(empty<GameResults>())
-            );
-        }),
-    ).subscribe(); // this sub is fine to live forever.  Once this is done, our
-    // server is done.
+  server
+    .on()
+    .pipe(mergeMap((sockets) => setupWithRxJS(sockets)))
```

Comment:
I left you a `mergeMap`. I know you like mergeMaps.

If this is still a performance issue, the entire thing could indeed be refactored to be a single, inline observable. But that seemed silly.